cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -0,0 +1,207 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/dataset/type_fwd.h"
+#include "arrow/dataset/visibility.h"
+#include "arrow/record_batch.h"
+#include "arrow/scalar.h"
+#include "arrow/type.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/stl.h"
+
+namespace arrow {
+namespace dataset {
+
+/// \brief Project a RecordBatch to a given schema.
+///
+/// Columns will be reordered to match the given schema.
+///
+/// Columns present in the given schema but absent from a record batch will be added as
+/// null or as arrays of a constant value.
+class RecordBatchProjector {
+ public:
+  RecordBatchProjector(MemoryPool* pool, std::shared_ptr<Schema> from,
+                       std::shared_ptr<Schema> to)
+      : pool_(pool),
+        from_(std::move(from)),
+        to_(std::move(to)),
+        missing_columns_(to_->num_fields(), nullptr),
+        column_indices_(to_->num_fields(), kNoMatch),
+        scalars_(to_->num_fields(), nullptr) {}
+
+  RecordBatchProjector(MemoryPool* pool, std::shared_ptr<Schema> from,
+                       std::shared_ptr<Schema> to,
+                       std::vector<std::shared_ptr<Scalar>> scalars)
+      : pool_(pool),
+        from_(std::move(from)),
+        to_(std::move(to)),
+        missing_columns_(to_->num_fields(), nullptr),
+        column_indices_(to_->num_fields(), kNoMatch),
+        scalars_(std::move(scalars)) {
+    DCHECK_EQ(scalars_.size(), missing_columns_.size());
+  }
+
+  Status Init() {
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      const auto& field = to_->field(i);
+      int matching_index = from_->GetFieldIndex(field->name());
+      if (matching_index != kNoMatch) {
+        if (!from_->field(matching_index)->Equals(field)) {
+          return Status::Invalid("fields had matching names but were not equivalent ",
+                                 from_->field(matching_index)->ToString(), " vs ",
+                                 field->ToString());
+        }
+      } else {
+        auto out_type = to_->field(i)->type();
+        RETURN_NOT_OK(MakeArrayOfNull(pool_, out_type, 0, &missing_columns_[i]));
```

Comment:
What happened to `scalars_` ?

Reply:
This is zero length for now; it's just a placeholder

Reply:
I'll add a comment